hack/lib/util/docs.sh

```diff
@@ -0,0 +1,139 @@
+  while read file; do
+    if [[ -e "${tmpdir}/${file}" && -n "${skipprefix}" ]]; then
+      local original generated
+      original=$(grep -v "^${skipprefix}" "${dest}/${file}") || :
+      generated=$(grep -v "^${skipprefix}" "${tmpdir}/${file}") || :
+      if [[ "${original}" == "${generated}" ]]; then
+        # overwrite generated with original.
+        mv "${dest}/${file}" "${tmpdir}/${file}"
+      fi
+    else
+      rm "${dest}/${file}" || true
+    fi
+  done <"${dest}/${filestore}"
+
+  # put the new generated file into the destination
+  find "${tmpdir}" -exec rsync -pt {} "${dest}" \; >/dev/null
+  #cleanup
+  rm -rf "${tmpdir}"
+
+  echo "Assets generated in ${dest}"
+}
+readonly -f generate_manual_pages
+
+function generate_documentation() {
+  local dest="$1"
+  local skipprefix="${1:-}"
+
+  # We do this in a tmpdir in case the dest has other non-autogenned files
+  # We don't want to include them in the list of gen'd files
+  local tmpdir="${OS_ROOT}/_tmp/gen_doc"
+  mkdir -p "${tmpdir}"
+  # generate the new files
+  gendocs "${tmpdir}"
+  # create the list of generated files
+  ls "${tmpdir}" | LC_ALL=C sort > "${tmpdir}/.files_generated"
+
+  # remove all old generated file from the destination
+  while read file; do
+    if [[ -e "${tmpdir}/${file}" && -n "${skipprefix}" ]]; then
+      local original generated
+      original=$(grep -v "^${skipprefix}" "${dest}/${file}") || :
+      generated=$(grep -v "^${skipprefix}" "${tmpdir}/${file}") || :
+      if [[ "${original}" == "${generated}" ]]; then
+        # overwrite generated with original.
+        mv "${dest}/${file}" "${tmpdir}/${file}"
+      fi
+    else
+      rm "${dest}/${file}" || true
+    fi
+  done <"${dest}/.files_generated"
+
+  # put the new generated file into the destination
+  find "${tmpdir}" -exec rsync -pt {} "${dest}" \; >/dev/null
+  #cleanup
+  rm -rf "${tmpdir}"
+
+  echo "Assets generated in ${dest}"
+}
+readonly -f generate_documentation
+
+# os::util::gen-docs generates docs and manpages for the all the binaries
+# created for Origin.
+function os::util::gen-docs() {
+  os::util::ensure::built_binary_exists 'gendocs'
+  os::util::ensure::built_binary_exists 'genman'
+
+  OUTPUT_DIR_REL=${1:-""}
+  OUTPUT_DIR="${OS_ROOT}/${OUTPUT_DIR_REL}/docs/generated"
+  MAN_OUTPUT_DIR="${OS_ROOT}/${OUTPUT_DIR_REL}/docs/man/man1"
+
+  mkdir -p "${OUTPUT_DIR}" || echo $? > /dev/null
+  mkdir -p "${MAN_OUTPUT_DIR}" || echo $? > /dev/null
+
+  generate_documentation "${OUTPUT_DIR}"
+  generate_manual_pages "${MAN_OUTPUT_DIR}" "oc"
+  generate_manual_pages "${MAN_OUTPUT_DIR}" "openshift"
+  generate_manual_pages "${MAN_OUTPUT_DIR}" "oadm"
+}
+readonly -f os::util::gen-docs
+
+# os::util::set-man-placeholder puts a placeholder for every generated manpage.
+function os::util::set-man-placeholder() {
+  MAN_OUTPUT_DIR="$1"
+  declare -A generated_files=( ["${1}/.files_generated_oadm"]="${1}/.files_generated_oadm" \
```

Comment:
> Not sure how to keep this cleaner other than using an associative array...

Maybe a function with a nice `if/elif` block? :)

> How early is "earlier versions of Bash used on OSX"? Is there a requirement for our scripts to work across all supported distros x versions of Bash?

Some people last I checked were running Bash 3, released in 2004. I think our support statement is that we need all members of the team with MacOS to be able to run these scripts. In the past I have been forbidden from using associative arrays for this reason. However, talking to some of the members recently, it looks like some of the newer OSX releases have very new versions of Bash (4.3) so they would be fine. Some part of me says let's use them and when someone hits a problem we can work around it then :)